walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,82 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<String> relevantAttendees = new ArrayList<>();
+    request.getAttendees().forEach(attendee -> relevantAttendees.add(attendee));
+
+    List<TimeRange> affectedTimeRanges = new ArrayList<>(); 
+    // get all affected time ranges
+    for (Event event : events) {
+      Set<String> scheduledAttendees = event.getAttendees(); 
+      for (String attendee : relevantAttendees) {
+        if (scheduledAttendees.contains(attendee)) {
+          // this event affects the final result
+          affectedTimeRanges.add(event.getWhen());
+          break;
+        }
+      }
+    }
+
+    long minRequiredDuration = request.getDuration();
+    long maxDuration = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true).duration();
+
+    List<TimeRange> availableRanges = new ArrayList<>();
+    if (minRequiredDuration > maxDuration) {
+      return availableRanges;
+    } else if (affectedTimeRanges.size() == 0) {
+      availableRanges.add(TimeRange.WHOLE_DAY);
+      return availableRanges;
+    }
+
+    // sort by start time
+    Collections.sort(affectedTimeRanges, TimeRange.ORDER_BY_START);
+
+    // consider the possibility of a range that starts at START
+    TimeRange firstAffectedRange = affectedTimeRanges.get(0); // no error because there is at least one time range
+    TimeRange potentialFirstRange = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, 
+        firstAffectedRange.start(), false);
+    if (potentialFirstRange.duration() >= minRequiredDuration) {
+      availableRanges.add(potentialFirstRange);
+    }
+    
+    // general case
+    TimeRange endsLastSoFar = affectedTimeRanges.get(0); 
+    for (TimeRange currRange: affectedTimeRanges) {
+      if (currRange.overlaps(endsLastSoFar)) {
+        // choose the one that ends later
+        endsLastSoFar = endsLastSoFar.end() > currRange.end() ? endsLastSoFar : currRange;
+        continue;
+      } 
+
+      // possibly found new available slot
+      int rangeStartTime = endsLastSoFar.end();
+      int rangeEndTime = currRange.start();
+      TimeRange potentialRange = TimeRange.fromStartEnd(rangeStartTime, rangeEndTime, false);
+      endsLastSoFar = currRange;
+
+      if (potentialRange.duration() < minRequiredDuration) {
+        continue; // duration is too short
+      }
+      availableRanges.add(potentialRange);
+    }
+
+    // consider the possibility of a range that ends at END
+    Collections.sort(affectedTimeRanges, TimeRange.ORDER_BY_END);
```

Comment:
Why do we need to sort it again? Should your `endsLastSoFar` be already at the last end range after the for loop above?

Reply:
Oh I forgot about that variable. 😂 Thanks!